### ISSUE TITLE
CodeIgniter Bypass

### ISSUE THREAD

**Issue by soaj1664** [bcit-ci/CodeIgniter#2667]
Hi Guys!

I had a chance to test CodeIgniter against XSS attack. I have found few things that might be interested for you guys:

I have created a test-bed here: http://xssplayground.net23.net/clean10.html

The area is protected by CodeIgniter but for testing purpose OUTPUT reflects in Standard HTML, Script, Attribute and URL context. In the following manner, in my test-bed output reflects in different context.

Output reflects here (Standard HTML Context): `<?php echo cleanCode($_POST['name']); ?> <br><br>`

Output reflects in attribute context `<div class='<?php echo cleanCode($_POST['name']);?>'>I am an attribute context</div> <br><br>`

Output reflects in URL context `<a href='<?php echo cleanCode($_POST['name']);?>'>I am URL context</a> <br><br>`

Output reflects in script context `<script> var a = '<?php echo cleanCode($_POST['name']);?>'; </script>`

The following injection will BYPASS CodeIgniter in "attribute" and "URL" context:

'; onmouseover=confirm(location);//

OR

'; onmouseover=confirm(cookie);//

In CodeIgniter's black-list you guys are checking against "document.cookie" and "window.location". The same purpose can be achieved by using cookie and location word! In above cases, if I inject 

'; onmouseover=alert(window.location);//

I got the following output which stops XSS:

Attribute context: `<div class='\'; onmouseover=alert&#40;[removed]&#41;;//'>`
URL context: `<a href='\'; onmouseover=alert&#40;[removed]&#41;;//'>`

Also in Chrome, the following injection bypassed URL context only:

j&NewLine;a&NewLine;vas&NewLine;cript:confirm(1);

If I simply inject javascript:alert(1) then CodeIgniter works perfectly and removed javascript but in the above manner, injection works.

Let me know, if I missed something :)

Regards

@soaj1664ashar

**Comment by RonnieSkansing** [comment]
As 2.x is open to XSS as described in this issue, I think it would be responsible to mention that is has known vuln/holes in the docs, as users are blindly trusting it atm and should not.

**Comment by narfbg** [comment]
@RonnieSkansing We've just released version 2.2.1 with these and some other security fixes.

I'm closing this issue now that it's addressed in both version trees and it got hard to follow anyway. Thank you all.
